### PR TITLE
Add OptionParser integration

### PR DESCRIPTION
**Pull request by jastkand**
@palkan this PR addresses this feature #17. Any comments and suggestions?

I've extracted the `serialize_val` method to `Anyway::Ext::ValueSerializer` module. Probably, it's better to extract it to an `Anyway::Utils` class. This way we won't have to include the module to the `Env` and `Config` classes and call it using the `Anyway::Utils.serialize_val(value)`. What do you think?

### CODE REVIEW

File: lib/anyway/ext/value_serializer.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+module Anyway
+  module Ext
+    # Extend Object through refinements
```
Comment (palkan) [inline lib/anyway/ext/value_serializer.rb]: ?

Comment (palkan) [reply lib/anyway/ext/value_serializer.rb]: Btw, we can implement serializer through refinements (as String extension, i.e. `val.serialize`).

File: lib/anyway/config.rb [hunk 2]
```diff
@@ -135,6 +158,18 @@ def load_from_env(config)
       config
     end
 
+    def option_parser
+      build_option_parser if @option_parser.nil?
```
Comment (palkan) [inline lib/anyway/config.rb]: Let's make it simpler: `@option_parser ||= build_option_parser`

File: lib/anyway/config.rb [hunk 3]
```diff
@@ -135,6 +158,18 @@ def load_from_env(config)
...
+      @option_parser
+    end
+
+    def parse_options!(options)
+      option_parser.parse!(options)
+    rescue OptionParser::InvalidOption
+      # NOTE: Do not fail when unknown arguments were given
```
Comment (palkan) [inline lib/anyway/config.rb]: Why do we want to ignore unknown arguments?

File: lib/anyway/config.rb [hunk 4]
```diff
@@ -155,5 +190,24 @@ def parse_yml(path)
         YAML.load_file(path)
       end
     end
+
+    def build_option_parser
+      require 'optparse'
+      @option_parser ||= OptionParser.new do |opts|
+        self.class.option_parser_attributes.each do |(key, options)|
```
Comment (palkan) [inline lib/anyway/config.rb]: What is `options`? Isn't `option_parser_attributes` an array of key?

File: lib/anyway/config.rb [hunk 5]
```diff
@@ -36,6 +37,26 @@ def config_name(val = nil)
...
 
+      def parse_options(*args, **hargs)
+        @option_parser_descriptions ||= {}
+        @option_parser_attributes ||= []
+
+        new_descriptions = hargs.deep_dup
+        new_descriptions.stringify_keys!
+        @option_parser_descriptions.merge! new_descriptions
+
+        new_keys = (args + new_descriptions.keys) - @option_parser_attributes
```
Comment (palkan) [inline lib/anyway/config.rb]: `@option_parser_attributes = @option_parser_attributes | (args + new_descriptions.keys)`?

File: lib/anyway/config.rb [hunk 6]
```diff
@@ -155,5 +190,24 @@ def parse_yml(path)
         YAML.load_file(path)
       end
     end
+
+    def build_option_parser
```
Comment (palkan) [inline lib/anyway/config.rb]: What about extracting this functionality into its own module?
For example, `OptionParserBuilder.call(options)` where `options` is a Hash of options names as keys and descriptions (or `nil`-s) as values?

Then we only need to generate a proper `options` Hash in the config itself.

That would make testing a little bit easier.

**Comment by jastkand** [comment]
@palkan Thanks for your response! I like the blacklisting API. The `ignore_options` and `describe_options` methods will make the usage of that feature more explicit and flexible.

> Another idea is to add DSL to extend parser (instead of overriding build_option_parser)

Sounds good!

> NOTE: we should also take into account inheritance: how do we want to handle inherited parsers attributes/extensions?

I think should behave the same way as `attr_config` works now (As I see the inherited class extends the configuration of parent class). Otherwise, it will be confusing for gem users that different methods are working differently.

P. S. I'll go through the comments later.

File: lib/anyway/ext/value_serializer.rb (hunk 1 above)
Comment (jastkand) [reply lib/anyway/ext/value_serializer.rb]: Copy-paste issue. Yes, I'll update the implementation to use refinements.

File: lib/anyway/config.rb (hunk 3 above)
Comment (jastkand) [reply lib/anyway/config.rb]: When unknown arguments are passes to `parse_options!` method, like:

```ruby
class MyConfig << Anyway::Config
  attr_config :host, :port
end

...

config.parse_option!(%w[--host localhost --port 3000 --log-level debug])
```

the mehtod call will raise an `OptionParser::InvalidOption` exception with the message "unknown argument '--log-level'"`. I don't think it's a good behavior to raise an exception in this case.

File: lib/anyway/config.rb (hunk 4 above)
Comment (jastkand) [reply lib/anyway/config.rb]: It's a leftover from the previous implementation. Initially, I wanted to store the `option_parser_attributes` as a hash, like
```ruby
{
  host: { description: 'Host name' },
  ...
}
```

Then I've changed my mind but didn't update the code accordingly.

File: lib/anyway/config.rb (hunk 6 above)
Comment (jastkand) [reply lib/anyway/config.rb]: Sounds good

**Comment by palkan** [comment]
> I think should behave the same way as attr_config works now (As I see the inherited class extends the configuration of parent class)

Looks like a bug :) `config_attributes` are not inherited: we initialize a new array for attribute [here](https://github.com/palkan/anyway_config/blob/master/lib/anyway/config.rb#L21) for every subclass.
Their should be smth like:

```ruby
@config_attributes ||= superclass <= Anyway::Config ? superclass.config_attributes.dup : []
```

The only thing that's  _inherited_ is `attr_accessor`.

File: lib/anyway/ext/hash.rb [hunk 7]
```diff
@@ -27,6 +27,8 @@ def stringify_keys!
             value.stringify_keys! if value.is_a?(::Hash)
             self[key.to_s] = value
           end
+
+          self
```
Comment (jastkand) [inline lib/anyway/ext/hash.rb]: Probably, it's better to extract this bug-fix in a separate PR?

**Comment by jastkand** [comment]
Hey, @palkan. It's ready for review.

File: lib/anyway/config.rb (hunk 3 above)
Comment (palkan) [reply lib/anyway/config.rb]: @jastkand  What about this one?

Comment (jastkand) [reply lib/anyway/config.rb]: > If a developers don't want to raise exceptions they could rescues them themselves

That's fair.

Anyway, let me confirm if I correctly understand the behavior which you expect. Let's say we have a class:

```ruby
class MyConfig < Anyway::Config
  attr_config :host, :port, :log_level

  ignore_options :log_level
end
```

Should it raise the `OptionParser::InvalidOption` exception when I call:

```ruby
config.parse_option!(%w[--host localhost --port 3000 --log-level debug])
```

Comment (palkan) [reply lib/anyway/config.rb]: Yep, it should raise an exception

**Comment by jastkand** [comment]
@palkan, I've updated a readme and changelog and removed the rescue from `parse_options!` method.

**Comment by palkan** [comment]
Thanks for your help!

**Comment by palkan** [comment]
Hey, @jastkand!
We're collecting contact information from the CultOfMartians participants (to send some goodies to you 🙂). Could you, please, send me your actual email address to `palkan@evl.ms`?

**Comment by jastkand** [comment]
Hey, @palkan! I've sent you my email address 🙂

**Pull request merged**